File: app/src/main/java/com/faltenreich/diaguard/shared/data/database/entity/Tag.java
package com.faltenreich.diaguard.shared.data.database.entity;

import androidx.room.ColumnInfo;
import androidx.room.Entity;

import com.faltenreich.diaguard.feature.export.Backupable;

@Entity
public class Tag extends BaseEntity implements Backupable {

    public static final String BACKUP_KEY = "tag";

    public class Column extends BaseEntity.Column {
        public static final String NAME = "name";
    }

    @ColumnInfo(name = Column.NAME)
    private String name;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    @Override
    public String getKeyForBackup() {
        return BACKUP_KEY;
    }

    @Override
    public String[] getValuesForBackup() {
        return new String[]{name};
    }

    public String getCharacter() {
        return name.substring(0, 1).toUpperCase();
    }
}

File: shared/src/commonMain/kotlin/com/faltenreich/diaguard/tag/Tag.kt
package com.faltenreich.diaguard.tag

import com.faltenreich.diaguard.datetime.DateTime
import com.faltenreich.diaguard.entry.Entry
import com.faltenreich.diaguard.shared.database.DatabaseEntity

/**
 * Entity for labeling an [Entry]
 */
sealed interface Tag {

    val name: String

    data class User(
        override val name: String,
    ) : Tag

    data class Local(
        override val id: Long,
        override val createdAt: DateTime,
        override val updatedAt: DateTime,
        override val name: String,
    ) : Tag, DatabaseEntity
}